### PR TITLE
Add specs to the API docs

### PR DESCRIPTION
**Pull request by duncanjbrown**
## Context

Following #4254 and #4329 we can build new API docs easily! The code that builds the docs lacks specs, though, and it's quite hard to follow. 

## Changes proposed in this pull request

- add specs for the public interfaces of the main classes `APIReference` and `APIOperation`
- refactor a bit to try to give `APIOperation` and `APISchema` (which `APIOperation` depends on) slightly clearer roles

## Guidance to review

Commit by commit!

## Things to check

- [X] This code does not rely on migrations in the same Pull Request
- [X] If this code includes a migration adding or changing columns, it also backfills existing records for consistency
- [X] API release notes have been updated if necessary
- [X] This code does not rely on the addition/removal of Azure config environment variables in the same Pull Request
- [X] New environment variables have been [added to the Azure config](https://github.com/DFE-Digital/apply-for-teacher-training/blob/master/docs/environment-variables.md#azure-hosting-devops-pipeline)

### CODE REVIEW

**Review by cwrw — APPROVED** [review]
👍

**Comment by duncanjbrown** [comment]
Thanks @cwrw !
